librad/src/git/replication.rs

```diff
@@ -181,119 +224,248 @@ where
-    person: Person,
-    remote_peer: PeerId,
-) -> Result<(), Error> {
-    let urn: Urn = Reference::rad_id(Namespace::from(person.urn()))
-        .with_remote(remote_peer)
-        .try_into()
-        .expect("namespace is set");
+fn ensure_rad_id(storage: &Storage, urn: &Urn, tip: ext::Oid) -> Result<(), Error> {
+    identities::common::IdRef::from(urn)
+        .create(storage, tip)
+        .map_err(|e| Error::Store(e.into()))
+}
 
-    match identities::person::verify(storage, &urn)? {
-        None => Err(Error::MissingIdentity),
-        Some(person) => {
-            // Create `rad/id` here, if not exists
-            ensure_rad_id(storage, &urn, person.content_id)?;
+mod person {
+    use super::*;
+
+    #[allow(clippy::unit_arg)]
+    #[tracing::instrument(level = "trace", skip(storage), err)]
+    pub fn ensure_setup(
+        storage: &Storage,
+        person: Person,
+        remote_peer: PeerId,
+    ) -> Result<(), Error> {
+        let urn: Urn = Reference::rad_id(Namespace::from(person.urn()))
+            .with_remote(remote_peer)
+            .try_into()
+            .expect("namespace is set");
 
-            // Track all delegations
-            for key in person.into_inner().doc.delegations {
-                tracking::track(storage, &urn, PeerId::from(key))?;
-            }
+        match identities::person::verify(storage, &urn)? {
+            None => Err(Error::MissingIdentity),
+            Some(person) => {
+                // Create `rad/id` here, if not exists
+                ensure_rad_id(storage, &urn, person.content_id)?;
 
-            Ok(())
-        },
+                // Track all delegations
+                for key in person.into_inner().doc.delegations {
+                    tracking::track(storage, &urn, PeerId::from(key))?;
+                }
+
+                Ok(())
+            },
+        }
     }
 }
 
-#[tracing::instrument(level = "trace", skip(storage), err)]
-fn ensure_setup_as_project(
-    storage: &Storage,
-    proj: Project,
-    remote_peer: PeerId,
-) -> Result<impl Iterator<Item = Urn>, Error> {
-    let local_peer_id = storage.peer_id();
+mod project {
+    use super::*;
 
-    let urn: Urn = Reference::rad_id(Namespace::from(proj.urn()))
-        .with_remote(remote_peer)
-        .try_into()
-        .expect("namespace is set");
+    #[derive(Clone, Debug)]
+    pub struct DelegateProject {
+        pub urn: Urn,
+        pub project: VerifiedProject,
+    }
 
-    // Verify + symref the delegates first
-    for delegate in proj.delegations().iter().indirect() {
-        let delegate_urn = delegate.urn();
-        // Find in `refs/namespaces/<urn>/refs/remotes/<remote
-        // peer>/rad/ids/<delegate.urn>`
-        let in_rad_ids: Urn = Reference::rad_delegate(Namespace::from(&urn), &delegate_urn)
+    pub fn verify(storage: &Storage, urn: Urn, remote: PeerId) -> Result<VerifiedProject, Error> {
+        let urn: Urn = Reference::rad_id(Namespace::from(urn.clone()))
```

Comment:
I think it would be nicer to have a separate function for the `Urn` conversion, without any side effect. The function could then have a doctest.